### PR TITLE
[NBEP] Type inference description

### PR DESCRIPTION
**Pull request by sklam**
A NBEP to describe the current type inference implementation.

(1st half of #2080)

### CODE REVIEW

File: docs/source/proposals/type-inference.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
+======================
+NBEP ?: Type Inference
+======================
+
+:Author: Siu Kwan Lam
+:Date: Sept 2015
```
Comment (seibert) [inline docs/source/proposals/type-inference.rst]: Sept 2016 :)

Comment (sklam) [inline docs/source/proposals/type-inference.rst]: it's a year long effort =P.

File: docs/source/proposals/type-inference.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  accepted for usability reason.
+
+* The *constraint network* is a dependency graph built from the IR.  Each
+  node represents an operation in the Numba IR and updates at least one type
+  variables.  There may be cycles due to loops in user code.
+
+The type inference process starts by seeding the argument types.  These initial
+types are propagated in the constraint network, which eventually fills all the
+type variables.  Due to cycles in the network, the process repeats until all
+type variables converge or it fails with undecidable types.
```
Comment (sklam) [inline docs/source/proposals/type-inference.rst]: should add why it terminates

Comment (sklam) [inline docs/source/proposals/type-inference.rst]: done

File: docs/source/proposals/type-inference.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+execution.  Upon an error, the type inference will set all types to the object
+type.  As a result, numba will fallback to *object-mode*.
+
+Since functions can be overloaded, the type inference needs to decide the call
+type, the calling signature, at each call site.  To decide the call type, the
+overload resolution is applied to all known versions of the callee function.
+These versions are described in *call-templates*.  A call-template can either be
+concrete or abstract.  A concrete call-template defines a fixed list of all
+possible signatures.  An abstract call-template defines the logic to compute
+the accepted signature.  It is used to implement generic functions.
```
Comment (stuartarchibald) [inline docs/source/proposals/type-inference.rst]: s/. It/ and it/

I assume the `It` refers to the abstract call-template?

File: docs/source/proposals/type-inference.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+Numba-compiled functions are generic functions due to their ability to compile
+new versions.  When it sees a new set of argument types, it triggers type
+inference to validate and determine the return type. When there are nested calls
+for numba-compiled functions, each call-site triggers type inference.
+This poses a problem to recursive functions because the type inference will also
+be triggered recursively.  Currently, simple single recursion is supported if
```
Comment (stuartarchibald) [inline docs/source/proposals/type-inference.rst]: You've fixed this :D

File: docs/source/proposals/type-inference.rst [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+flow need extra care.  Each incoming paths could be holding a different
+variable version.  A new version is implicitly created at the merge point.
+The incoming versions are merged.  This may translate to an implicit cast.
+
+Numba uses function overloading to emulate Python duck-typing.  The type of a
+function can contain multiple call signatures that accept different argument
+types and yields different return types.  The process to decide the best
+signature for an overloaded function is called *overload resolution*.
+Numba partially implements the C++ overload resolution scheme
+(ISOCPP 13.3 Overload Resolution).  The scheme uses a "best fit" algorithm by
```
Comment (sklam) [inline docs/source/proposals/type-inference.rst]: The official standard is not free.  I can point to the working draft http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4296.pdf

File: docs/source/proposals/type-inference.rst [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+A failure in type inference can be caused by two reasons.  One reason is user
+error due to incorrect use of a type.  This type of error will also trigger an
+exception in regular python execution.  Another reason is due to the use of an
+unsupported feature, but the code is otherwise valid in regular python
+execution.  Upon an error, the type inference will set all types to the object
+type.  As a result, numba will fallback to *object-mode*.
+
+Since functions can be overloaded, the type inference needs to decide the call
+type, the calling signature, at each call site.  To decide the call type, the
```
Comment (sklam) [inline docs/source/proposals/type-inference.rst]: I am going to remove  `call type` from the document and use `type signature`.

**Comment by stuartarchibald** [comment]
Thanks for the patches and doing the fixes. This document is a great addition. Merging!

**Pull request merged**